posts/wit-bringing-types-to-near-contracts.md

```diff
@@ -0,0 +1,310 @@
+Assume that you have a Rust smart contract:
+
+```rust
+use near_sdk::{witgen, near_bindgen}
+
+/// A message that contains some text
+#[witgen]
+pub struct Message {
+  /// Inner string value
+  text: String,
+}
+
+//...
+pub struct Contract {
+  message: Message
+}
+
+#[near_bindgen]
+impl Contract {
+  
+  /// A change call to set the message
+  pub fn set_message(&mut self, message: Message) {
+    self.mesage = message;
+  }
+
+  /// A view call to get the current message
+  pub fn get_message(self) -> Message {
+    self.message
+  }
+}
+
+```
+
+The view method `get_message` returns a `Message` struct. `#[witgen]` is a Rust macro that generates a corresponding [`wit` record](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md#item-record-bag-of-named-fields). See the [witgen repo](https://github.com/bnjjj/witgen)
+
+Furthermore the `#[near_bindgen]` macro has been updated to generate [function types in `wit`](https://github.com/bytecodealliance/wit-bindgen/blob/main/WIT.md#item-function).
+
+Next you can use `witme` to generate a `.wit` file for the contract.
+
+```bash
+witme near wit
+```
+
+generates `index.wit`
+
+```wit
+///  A message that contains some text
+record message {
+    ///  Inner string value
+    text: string
+}
+
+///  A change call to set the message
+///  change
+set-message: function(message: message)
+
+///  A view call to get the current message
+get-message: function() -> message
+
+```
+
+This `wit` file now describes the Contract's interface in a language agnostic way. This means we can now generate source code for a different language.  For example, TypeScript:
+
+```bash
+witme near ts
+```
+
+By default this command looks for an `index.wit` and puts the generated TS in it's own `./ts` folder.
+
+```typescript
+/**
+* A message that contains some text
+*/
+export interface Message {
+  /**
+  * Inner string value
+  */
+  text: string;
+}
+
+export class Contract {
+  /** Account calling the contract and the contractId to call */
+  constructor(public account: Account, public readonly contractId: string){}
+  
+  /**
+  * A change call to set the message
+  */
+  async set_message(args: {
+    message: Message;
+  }, options?: ChangeMethodOptions): Promise<void> {}
+
+  /**
+  * A view call to get the current message
+  */
+  get_message(args = {}, options?: ViewFunctionOptions): Promise<Message> {}
+}
+
+```
+
+This means the contract's interface is now available to use with `near-api-js` to interact with the contract.
+
+```typescript
+
+import {Contract, Message} from "message/contract";
+import {Account} from "near-api-js";
+
+async function getMessage(currentAccount: Account): Promise<Message> {
+  let contract = new Contract(currentAccount, "contract.testnet");
+  return contract.get_message();
+}
+
+
+```
+
+Since the original comments in the rust code are preserved you can also get hover over docs in your IDE, or generate the docs see [TenK's docs](https://tenk-dao.github.io/tenk/docs/).
```

Comment:
```suggestion
Since the original comments in the rust code are preserved you can also get hover over docs in your IDE, or generate a documentation website (see [TenK's docs](https://tenk-dao.github.io/tenk/docs/) for an example).
```